### PR TITLE
[Merged by Bors] - Refactor: use `Option.map₂`

### PR DESCRIPTION
**Pull request by urkud**
This is a forward-port of leanprover-community/mathlib#18081

### CODE REVIEW

**Comment by kbuzzard** [comment]
In `Mathlib/Algebra/Group/WithOne/Defs.lean`, the `show`s can be removed in `WithOne.mulOneClass` (in your PR you change them to `id`s -- in mathlib4 they can be removed completely).
```lean
instance mulOneClass [Mul α] : MulOneClass (WithOne α) where
  mul := (· * ·)
  one := 1
  one_mul := (Option.liftOrGet_isLeftId _).1
  mul_one := (Option.liftOrGet_isRightId _).1
```

File: Mathlib/Algebra/Group/WithOne/Defs.lean [hunk 1]
```diff
@@ -226,57 +226,33 @@ theorem coe_one [One α] : ((1 : α) : WithZero α) = 1 :=
...
   { WithZero.mulZeroClass, WithZero.one with
-    one_mul := fun a =>
-      match a with
-      | none => rfl
-      | some a => congr_arg some <| one_mul a,
-    mul_one := fun a =>
-      match a with
-      | none => rfl
-      | some a => congr_arg some <| mul_one a }
+    -- In Lean 3, without the `id`, the generated `_proof_1` lemma has the wrong type
```
Comment (kbuzzard) [inline Mathlib/Algebra/Group/WithOne/Defs.lean:253]: This is also true in Lean 4.
```suggestion
    -- Without the `id`, the generated `_proof_1` lemma has the wrong type
```

**Comment by kbuzzard** [comment]
Algebra.Order.Ring.WithTop: in mathlib3 you removed the analogue of
```lean
instance noZeroDivisors [MulZeroClass α] [NoZeroDivisors α] : NoZeroDivisors (WithTop α) :=
```
(it's on line 154) and it should be removed here too I guess.

**Review by kbuzzard — COMMENTED** [review]
I've ploughed through this to check that the PRs do coincide. I left two comments (above) about when they didn't. Modulo these I approve!

**Comment by urkud** [comment]
@kbuzzard Thanks a lot for the review!

**Comment by fpvandoorn** [comment]
Please merge master and double check that no new files of your PR have been ported.

bors d+

**Comment by urkud** [comment]
bors merge

**Pull request closed without merging**